tests/test-grammar-integration.cpp

```diff
@@ -130,75 +132,161 @@ ws ::= [ \t\n\r]?)""";
+        bool matched = match_string(test_string, grammar);
 
-        bool completed_grammar = false;
+        assert(!matched);
 
-        for (const auto & stack : grammar->stacks) {
-            if (stack.empty()) {
-                completed_grammar = true;
-                break;
+        // Reset the grammar stacks
+        grammar->stacks = original_stacks;
+    }
+
+    // Clean up allocated memory
+    llama_grammar_free(grammar);
+}
+
+static void test_quantifiers() {
+    // Populate test data with grammar strings and their associated collections of expected passing and failing strings
+    const std::vector<
+        std::tuple<
+            std::string,
+            std::vector<std::string>,
+            std::vector<std::string>>>
+        test_data = {
+        {
+            // Grammar
+            R"""(root ::= "a"*)""",
+            // Passing strings
+            {
+                "",
+                "a",
+                "aaaaa",
+                "aaaaaaaaaaaaaaaaaa",
+                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+            },
+            // Failing strings
+            {
+                "b",
+                "ab",
+                "aab",
+                "ba",
+                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
+            }
+        },
+        {
+            // Grammar
+            R"""(root ::= "a"+)""",
+            // Passing strings
+            {
+                "a",
+                "aaaaa",
+                "aaaaaaaaaaaaaaaaaa",
+                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+            },
+            // Failing strings
+            {
+                "",
+                "b",
+                "ab",
+                "aab",
+                "ba",
+                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
+            }
+        },
+        {
+            // Grammar
+            R"""(root ::= "a"?)""",
+            // Passing strings
+            {
+                "",
+                "a"
+            },
+            // Failing strings
+            {
+                "b",
+                "ab",
+                "aa",
+                "ba",
+            }
+        },
+        {
+            // Grammar
+            R"""(
+root ::= cons+ vowel* cons? (vowel cons)*
+vowel ::= [aeiouy]
+cons ::= [bcdfghjklmnpqrstvwxyz]
+)""",
+            // Passing strings
+            {
+                "yes",
+                "no",
+                "noyes",
+                "crwth",
+                "four",
+                "bryyyy",
+            },
+            // Failing strings
+            {
+                "yess",
+                "yesno",
+                "forty",
+                "catyyy",
             }
         }
+    };
 
-        assert(completed_grammar);
+    for (const auto & test_datum : test_data) {
+        const auto & grammar_str = std::get<0>(test_datum);
```

Comment:
Maybe print this? ( and then slow-drip print each passing / failing string being tested, on one or two lines)

I personally like verbose tests, this would make it easier to understand what caused any failure.

Reply:
Note that I do print this in the event of a failure (c.f. line 263), but I also don't hate verbose tests, and I'll see about adding these prints in by default.